aten/src/ATen/native/SpatialAveragePooling.cpp

```diff
@@ -0,0 +1,465 @@
+          int64_t wend = std::min(wstart + kW, inputWidth + padW);
+          int pool_size = (hend - hstart) * (wend - wstart);
+          hstart = std::max(hstart, (int64_t) 0);
+          wstart = std::max(wstart, (int64_t) 0);
+          hend = std::min(hend, inputHeight);
+          wend = std::min(wend, inputWidth);
+
+          scalar_t z = *ptr_gradOutput++;
+
+          int divide_factor;
+          if(count_include_pad)
+            divide_factor = pool_size;
+          else
+            divide_factor = (hend - hstart) * (wend - wstart);
+
+          int64_t kx, ky;
+          for(ky = hstart ; ky < hend; ky++)
+          {
+            for(kx = wstart; kx < wend; kx++)
+              ptr_gradInput[ky*inputWidth + kx] += z/divide_factor;
+          }
+        }
+      }
+    }
+  }
+}
+
+static void avg_pool2d_backward_cpu_template(
+  Tensor& input,
+  Tensor& gradOutput,
+  Tensor& gradInput,
+  IntList kernel_size,
+  IntList stride_size,
+  IntList pad_size,
+  bool ceil_mode,
+  bool count_include_pad)
+{
+  auto kH = kernel_size[0];
+  auto kW = kernel_size[1];
+  auto dH = stride_size[0];
+  auto dW = stride_size[1];
+  auto padH = pad_size[0];
+  auto padW = pad_size[1];
+  int dimw = 2;
+  int dimh = 1;
+  int dimc = 0;
+  int64_t nbatch = 1;
+  int64_t ndim = 3;
+
+  int64_t inputWidth;
+  int64_t inputHeight;
+  int64_t outputWidth;
+  int64_t outputHeight;
+  int64_t nInputPlane; // number of channels (or colors)
+
+  avg_pool2d_shapecheck(
+      input, &gradOutput, kernel_size, stride_size, pad_size, ceil_mode);
+
+
+  if (input.dim() == 4) {
+    nbatch = input.size(0);
+    dimw++;
+    dimh++;
+    dimc++;
+    ndim = 4;
+  }
+
+  inputWidth = input.size(dimw);
+  inputHeight = input.size(dimh);
+  nInputPlane = input.size(dimc);
+
+  outputWidth = pooling_output_shape<int64_t>(inputWidth, kW, padW, dW, 1, ceil_mode);
+  outputHeight = pooling_output_shape<int64_t>(inputHeight, kH, padH, dH, 1, ceil_mode);
+
+  gradInput.resize_as_(input);
+
+  if (input.dim() == 3)
+    gradOutput.resize_({nInputPlane, outputHeight, outputWidth});
+  else
+    gradOutput.resize_({input.size(0), nInputPlane, outputHeight, outputWidth});
+
+  gradOutput = gradOutput.contiguous();
+  if (!gradInput.is_contiguous())
+    AT_ERROR("gradInput must be contiguous");
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    gradOutput.scalar_type(), "avg_pool2d", [&] {
+      avg_pool2d_backward_cpu_frame<scalar_t>(
+          gradInput,
+          gradOutput,
+          nInputPlane,
+          inputHeight,
+          inputWidth,
+          outputHeight,
+          outputWidth,
+          nbatch,
+          count_include_pad,
+          kernel_size,
+          stride_size,
+          pad_size);
+    });
+}
+
+} // namespace
+
+Tensor& avg_pool2d_out_cpu(
+  Tensor& input,
+  IntList kernel_size,
+  IntList stride,
+  IntList padding,
+  bool ceil_mode,
+  bool count_include_pad,
+  Tensor& output)
```

Comment:
These signatures need to match the generated ones in `build/aten/src/ATen/NativeFunctions.h`. The `out` argument is moved to the front in that header and does not necessarily match the order in `native_functions.yaml`.

Reply:
I'm pretty sure this is why you get build failures.